projects/libyaml/libyaml_emitter_fuzzer.c

```diff
@@ -0,0 +1,322 @@
+    return false;
+
+  default:
+    return false;
+  }
+}
+
+bool copy_event(yaml_event_t *event_to, yaml_event_t *event_from) {
+
+  switch (event_from->type) {
+  case YAML_STREAM_START_EVENT:
+    return yaml_stream_start_event_initialize(
+        event_to, event_from->data.stream_start.encoding);
+
+  case YAML_STREAM_END_EVENT:
+    return yaml_stream_end_event_initialize(event_to);
+
+  case YAML_DOCUMENT_START_EVENT:
+    return yaml_document_start_event_initialize(
+        event_to, event_from->data.document_start.version_directive,
+        event_from->data.document_start.tag_directives.start,
+        event_from->data.document_start.tag_directives.end,
+        event_from->data.document_start.implicit);
+
+  case YAML_DOCUMENT_END_EVENT:
+    return yaml_document_end_event_initialize(
+        event_to, event_from->data.document_end.implicit);
+
+  case YAML_ALIAS_EVENT:
+    return yaml_alias_event_initialize(event_to, event_from->data.alias.anchor);
+
+  case YAML_SCALAR_EVENT:
+    return yaml_scalar_event_initialize(
+        event_to, event_from->data.scalar.anchor, event_from->data.scalar.tag,
+        event_from->data.scalar.value, event_from->data.scalar.length,
+        event_from->data.scalar.plain_implicit,
+        event_from->data.scalar.quoted_implicit, event_from->data.scalar.style);
+
+  case YAML_SEQUENCE_START_EVENT:
+    return yaml_sequence_start_event_initialize(
+        event_to, event_from->data.sequence_start.anchor,
+        event_from->data.sequence_start.tag,
+        event_from->data.sequence_start.implicit,
+        event_from->data.sequence_start.style);
+
+  case YAML_SEQUENCE_END_EVENT:
+    return yaml_sequence_end_event_initialize(event_to);
+
+  case YAML_MAPPING_START_EVENT:
+    return yaml_mapping_start_event_initialize(
+        event_to, event_from->data.mapping_start.anchor,
+        event_from->data.mapping_start.tag,
+        event_from->data.mapping_start.implicit,
+        event_from->data.mapping_start.style);
+
+  case YAML_MAPPING_END_EVENT:
+    return yaml_mapping_end_event_initialize(event_to);
+  }
+
+  return false;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  if (size < 2)
+    return 0;
+
+  yaml_parser_t parser;
+  yaml_emitter_t emitter;
+  yaml_event_t event;
+  unsigned char buffer[BUFFER_SIZE + 1];
+  size_t written = 0;
+  yaml_event_t events[MAX_EVENTS];
+  size_t event_number = 0;
+  bool done = false;
+  int count = 0;
+  int error = 0;
+  bool is_canonical = data[0] & 1;
+  bool is_unicode = data[1] & 1;
+  data = data + 2;
+  size = size - 2;
+
+  if (!yaml_parser_initialize(&parser))
+    return 0;
+
+  yaml_parser_set_input_string(&parser, data, size);
+  if (!yaml_emitter_initialize(&emitter)) {
+    yaml_parser_delete(&parser);
+    return 0;
+  }
+
+  yaml_emitter_set_canonical(&emitter, is_canonical);
+  yaml_emitter_set_unicode(&emitter, is_unicode);
+
+  yaml_emitter_set_output(&emitter, yaml_write_handler, &emitter);
```

Comment:
hmm, that does not make sense to me.
iiuc, you just implement the same behavior that existed before.
my intention was to remove arbitrary buffer sizes and instead have a dynamically growing heap buffer.

i would not manipulate the internal data structures in the emitter, things might break.

the implementation of the write handler could be similar to this:

have a
struct { char*begin; size_t size; };
create an instance of this(with {NULL, 0} ) before calling yaml_emitter_set_output and give a pointer to it as last arg.

now in the function:
- you malloc the pointer with the size needed when it's the first call (begin is NULL)
- realloc the buffer to the new required size
- memcpy the new data to buf->begin + buf->size and update size with the new value

you could make more complicated implementations, if you want.

Reply:
Ah, I see your point. Apologies for the confusion, thanks for clarifying.